### PR TITLE
Webhook polishing

### LINKED ISSUE #2276: Add secret signature to the webhook payload
### Is your proposal related to a problem?
Currently, the webhook doesn't have a signature that can be used to verify the source. Adding a secret signature would help in making the webhooks a lot more secure.  

### Describe the solution you'd like
Something similar to what Github does for its webhooks would be ideal.

### PR DESCRIPTION
**Pull request by alishaz-polymath**
## What does this PR do?

- [x] Adds BookingId as part of the webhook payload
- [x] Adds a secret signature to verify webhook's authenticity 
- [x] Webhooks are now triggered when the user requests reschedule

Fixes calcom/cal.com#2276 calcom/cal.com#2877 

[Loom Video](https://www.loom.com/share/a4d3e2c5be5d4c858f839810212b4e9a)

**Environment**: Staging(main branch)

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

## How should this be tested?

- Simply add a new secret key to your webhook and save. Next, wait for the webhook trigger (event created, cancelled or rescheduled, whatever is subscribed) and then use the secret key to create an hmac, and update that with the body of the webhook data to create an SHA256. Compare the hash received in the header of the webhook `(X-Cal-Signature-256)` with the one created using the secret key and the body of the payload.

### CODE REVIEW

File: apps/web/components/webhook/WebhookDialogForm.tsx [hunk 1]
```diff
@@ -29,20 +30,39 @@ export default function WebhookDialogForm(props: {
...
+  const hasSecretKey = !!defaultValues.secret;
+  const currentSecret = defaultValues.secret;
 
   const form = useForm({
     defaultValues,
   });
+
+  const handleInput = (event: React.FormEvent<HTMLInputElement>) => {
+    setNewSecret(event.currentTarget.value);
+  };
```
Comment (alishaz-polymath) [inline apps/web/components/webhook/WebhookDialogForm.tsx:49]: Handles controlled input to allow entering new secret key in the input

File: apps/web/components/webhook/WebhookDialogForm.tsx [hunk 2]
```diff
@@ -29,20 +30,39 @@ export default function WebhookDialogForm(props: {
...
+
+  useEffect(() => {
+    if (changeSecret) {
+      form.unregister("secret", { keepDefaultValue: false });
+    }
+  }, [changeSecret]);
```
Comment (alishaz-polymath) [inline apps/web/components/webhook/WebhookDialogForm.tsx:55]: Ensures that the currently using secret key never shows up to the user screen

File: apps/web/components/webhook/WebhookDialogForm.tsx [hunk 3]
```diff
@@ -29,20 +30,39 @@ export default function WebhookDialogForm(props: {
...
+
   return (
     <Form
       data-testid="WebhookDialogForm"
       form={form}
       handleSubmit={async (event) => {
-        const e = { ...event, eventTypeId: props.eventTypeId };
+        const e = changeSecret
+          ? { ...event, eventTypeId: props.eventTypeId }
+          : { ...event, secret: currentSecret, eventTypeId: props.eventTypeId };
```
Comment (alishaz-polymath) [inline apps/web/components/webhook/WebhookDialogForm.tsx:64]: Pass the old secret unless new one is set (or left blank  which suggests the user wishes to remove the secret key)

File: apps/web/lib/webhooks/sendPayload.tsx [hunk 4]
```diff
@@ -56,10 +59,15 @@ const sendPayload = async (
     });
   }
 
+  const secretSignature = secretKey
+    ? createHmac("sha256", secretKey).update(`${body}`).digest("hex")
+    : "no-secret-provided";
```
Comment (alishaz-polymath) [inline apps/web/lib/webhooks/sendPayload.tsx:64]: Creates a signature if secret exists, else `no-secret-provided`

File: apps/web/lib/webhooks/sendPayload.tsx [hunk 5]
```diff
@@ -56,10 +59,15 @@ const sendPayload = async (
...
+
   const response = await fetch(subscriberUrl, {
     method: "POST",
     headers: {
       "Content-Type": contentType,
+      "X-Cal-Signature-256": secretSignature,
```
Comment (alishaz-polymath) [inline apps/web/lib/webhooks/sendPayload.tsx:70]: Pass secret signature in the webhook call header

File: apps/web/pages/api/book/request-reschedule.ts [hunk 6]
```diff
@@ -216,6 +229,45 @@ const handler = async (
...
+      const webhooks = await getWebhooks(subscriberOptions);
+      const promises = webhooks.map((webhook) =>
+        sendPayload(webhook.secret, eventTrigger, new Date().toISOString(), webhook, evt).catch((e) => {
+          console.error(
+            `Error executing webhook for event: ${eventTrigger}, URL: ${webhook.subscriberUrl}`,
+            e
+          );
+        })
+      );
+      await Promise.all(promises);
```
Comment (alishaz-polymath) [inline apps/web/pages/api/book/request-reschedule.ts:270]: Add webhook when reschedule is requested

**Review by alishaz-polymath — COMMENTED** [review]
Left self-review comments (Loom also attached in the PR description)

File: apps/web/components/webhook/WebhookDialogForm.tsx [hunk 7]
```diff
@@ -119,6 +138,50 @@ export default function WebhookDialogForm(props: {
...
+              onClick={() => {
+                setChangeSecret(true);
+              }}>
+              {t("change_secret")}
+            </Button>
+          </>
+        )}
+        {!!hasSecretKey && changeSecret && (
+          <>
+            <TextField
```
Comment (emrysal) [inline apps/web/components/webhook/WebhookDialogForm.tsx:161]: I'd wrap this in a Controller
```typescript
            <Controller
              name="secret"
              render={({ field }) => <TextField {...field} />
```

**Review by emrysal — APPROVED** [review]

**Pull request merged**